### PR TITLE
Add Ascii Reader

### PR DESCRIPTION
**Pull request by fstephany**
As discussed in issue #42, here's a first pass on an ASCII reader.
I haven't used the fuzzer. If you have direction on how to setup and test it, I would be glad to add id.

### CODE REVIEW

File: src/stream/ascii_reader.rs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        }
+    }
+
+    fn error(&self, kind: ErrorKind) -> Error {
+        kind.with_byte_offset(self.current_pos)
+    }
+
+    /// Get a char without consuming it.
+    fn peek(&mut self) -> Option<u8> {
+        // consume a char then rollback to previous position.
```
Comment (ebarnard) [inline src/stream/ascii_reader.rs]: It would be nice to avoid the `Seek` bound on the reader.

How about storing a `peeked_byte: Option<u8>` in `AsciiReader` in `peek` and consuming it if it exists in `advance`?

**Comment by ebarnard** [comment]
The fuzzer is fairly straightforward, we use cargo-fuzz. If you follow the (minimal) [guide](https://rust-fuzz.github.io/book/cargo-fuzz.html) to get it installed and then create a new fuzz-target based on the `xml_reader.rs` target and run it that should just work. You can then put some valid ascii plists in the corpus folder to speed things up a bit if you want.

File: src/stream/ascii_reader.rs (hunk 1 above)
Comment (fstephany) [reply src/stream/ascii_reader.rs]: Oh you're right. That should be pretty simple to add and will indeed remove the `Seek` requirement.

File: src/stream/ascii_reader.rs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        let _ = self.reader.seek(SeekFrom::Current(-1));
+        self.update_current_pos();
+        peeked
+    }
+
+    /// Consume the reader and return the next char.
+    fn advance(&mut self) -> Option<u8> {
+        let mut buf: [u8; 1] = [0; 1];
+
+        match self.reader.read(&mut buf) {
```
[the remark this answers is not in the document]
Comment (fstephany) [reply src/stream/ascii_reader.rs]: Good catch. I'll update that.

**Comment by madig** [comment]
Woah! I am interested in this. I'm working in the font design space and the dominant font design application https://glyphsapp.com uses a "dialect" of the old ASCII plist format for primary storage, with some adjustment here and there "to save bytes". I.e. it will leave quotes out if something is probably a string. @anthrotype created a Python/Cython implementation at https://github.com/fonttools/openstep-plist that handles these.

I am going to play around with this PR and see how far I can get.

**Comment by fstephany** [comment]
@madig That's another use case indeed ;) 
If you have some sample files for testing, it could be nice to check that the code is covering all the cases. If you want to port some unit tests from the Python lib, go ahead!

@ebarnard Thanks a lot for the comments. If you see further things to change do not hesitate. Even if it is code style conformance.

I'm not that sure how to bubble the IO errors from `read_exact` so I've just added another `ErrorKind`. If you had something clever in mind, I'll change my naive implementation.

**Comment by madig** [comment]
@fstephany: On it. I have test files over at https://github.com/madig/glyphsrw-rs/tree/master/data that I'm feeding into my toy implementation. Currently trying to port over the the string-or-int-or-float logic from https://github.com/fonttools/openstep-plist/blob/master/src/openstep_plist/parser.pyx#L257-L299. What I have so far:

```Rust
enum UnquotedLiteralType {
    String,
    Integer,
    Float,
}

fn guess_unquoted_string_literal_type(buffer: &str) -> UnquotedLiteralType {
    let mut maybe_number = true;
    let mut is_float = false;
    let mut ch: char = buffer.chars().nth(0).unwrap();
    let mut i: usize = 0;

    match ch {
        '-' => {
            if buffer.len() > 1 {
                i += 1;
                ch = buffer.chars().nth(i).unwrap();
                if !ch.is_ascii_digit() {
                    maybe_number = false
                }
            } else {
                maybe_number = false
            }
        }
        other => {
            if !other.is_ascii_digit() {
                maybe_number = false
            }
        }
    }

    if maybe_number {
        for c in buffer[i..].bytes() {
            if c > ('9' as u8) || c < ('.' as u8) || c == ('/' as u8) {
                return UnquotedLiteralType::String; // Not a number.
            } else if c == ('.' as u8) {
                if !is_float {
                    is_float = true;
                } else {
                    // Seen a second '.', so it's not a float.
                    return UnquotedLiteralType::String;
                }
            }
        }

        if is_float {
            return UnquotedLiteralType::Float;
        } else {
            return UnquotedLiteralType::Integer;
        }
    }

    UnquotedLiteralType::String
}

[...]

        let string_literal = String::from_utf8(acc)
            .map_err(|_e| self.error(ErrorKind::InvalidUtf8AsciiStream))?;

        match guess_unquoted_string_literal_type(&string_literal) {
            UnquotedLiteralType::String => Ok(Some(Event::String(string_literal))),
            UnquotedLiteralType::Integer => {
                match Integer::from_str(&string_literal) {
                    Ok(i) => Ok(Some(Event::Integer(i))),
                    Err(_) => Err(self.error(ErrorKind::InvalidIntegerString)),
                }
            },
            UnquotedLiteralType::Float => {
                match string_literal.parse::<f64>() {
                    Ok(f) => Ok(Some(Event::Real(f))),
                    Err(_) => Err(self.error(ErrorKind::InvalidRealString)),
                }
            }
        }
```

**Comment by fstephany** [comment]
Oh, this is _much_ more involved than what I'm currently doing: 

~~~rust
match Integer::from_str(&string_literal) {
       Ok(i) => Ok(Some(Event::Integer(i))),
       Err(_) => Ok(Some(Event::String(string_literal))),
}
~~~

I just accumulate the characters and then try to convert the resulting string to an integer. I do not support floats (actually, even Integer are not part of the Apple spec).

**Comment by madig** [comment]
Fair enough. This may be overkill for anything not Glyphs.

**Comment by madig** [comment]
(If there is interest in upstreaming the more complicated logic, maybe this guessing should be optional and off by default since it is not part of the spec.)

**Comment by madig** [comment]
(Ugh, ignore my code above. Your code is more robust: either it parses as an int (or maybe a float) or it doesn't and then is a string)

**Comment by madig** [comment]
Hm. Strings with escaped quotes aren't handled properly right now, example:

```text
{
filter = "subCategory == \"Currency\"";
locked = 1;
name = currency;
position = "{-196, 760}";
}
```

**Comment by fstephany** [comment]
@madig I've fixed it. Can you check the simple test case to be sure that it was the expected behavior ?

That makes me thing that I should add unit tests for invalid inputs. I don't know if I should completely reject malformed files (e.g., missing comma in a list) or be tolerant.  I feel like this format has been abused by so many tools with their own conventions that it might be more effective to accept inputs even if they don't comply 100% with the format.

**Comment by madig** [comment]
Can do at some point this week. Yeah, I want to look into tests for this at some point. There's more stuff like escape sequences (think "hello\012world") and some tricky handling of Unicode escapes that openstep-plist handles.

File: src/stream/ascii_reader.rs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    Integer,
+};
+use std::io::Read;
+
+pub struct AsciiReader<R: Read> {
+    reader: R,
+    current_pos: u64,
+
+    /// lookahead char to avoid backtracking.
+    peeked_char: Option<u8>,
```
Comment (ebarnard) [inline src/stream/ascii_reader.rs]: You could even use `bytes: Peekable<Bytes<R>>` constructed from `reader.bytes().peekable()` here. Then you have:
```rust
fn advance(&mut self) -> Result<Option<u8>, Error> {
    // Might need to split this into two parts to make the borrow checker happy.
    let ret = self.bytes.next().transpose().map_err(|err| self.error(ErrorKind::Io(err)));
    self.current_pos += 1;
    ret
}

fn peek(&mut self) -> Result<Option<u8>, Error> {
    // As before WRT the borrow checker.
    self.bytes.peek().transpose().map_err(|err| self.error(ErrorKind::Io(err)))
}
```

(BTW no need whatsoever to make this change, just thought it was neat)

**Comment by ebarnard** [comment]
# Dialects

This library aims to be minimal and focussed on parsing Apple spec Plists. This means that a Plist that follows Apple's spec should always parse per the spec.

I'm happy to support dialects/extensions so long as they do not affect the parsing of Apple spec Plists and have a simple implementation.

Extensions that require complex code to implement are, unfortunately, out of scope for this library. Any extension that requires a configuration option falls into this category.

@madig I'd be happy to support the dialect you've made for glyphsapp provided we can get it parsing without configuration options.

# Malformed Files

@fstephany I'm not keen to make an effort to parse obviously malformed files as I don't want people to rely on how this library interprets them, which could change as bugs are fixed or new extensions added. I'd rather we work to explicitly support various common dialects (per the limitations above).

File: src/stream/ascii_reader.rs [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+        return Ok(None);
+    }
+}
+
+impl<R: Read> Iterator for AsciiReader<R> {
+    type Item = Result<Event, Error>;
+
+    fn next(&mut self) -> Option<Result<Event, Error>> {
+        match self.read_next() {
```
Comment (ebarnard) [inline src/stream/ascii_reader.rs]: This can be just `self.read_next().transpose()`. Don't worry that this will fail on the 1.32.0 Travis builder.

**Comment by madig** [comment]
@ebarnard thanks for the feedback. I understand that if I wanted any crazy parsing, I'd need to make my own library (I did not come up with that dialect, I just need to parse it :grin:). I think the code should be able to handle the Glyphs dialect by simply not interpreting strings as anything and instead leaving that to client code. I think that is how the specification intends it. What may be needed is slightly more complex parsing of strings like marked in https://github.com/fonttools/openstep-plist/blob/master/tests/test_parser.py#L135-L152. I can see if I can implement that when I get some time. Maybe a fuzzer could find even more corner cases :thinking:

**Comment by ebarnard** [comment]
@madig Those examples all look to be supported by Apple's parser as well: https://pewpewthespells.com/blog/dangers_of_ascii_plists.html. It's probably worth reading the whole of [CFOldStylePlist.c](http://www.opensource.apple.com/source/CF/CF-1153.18/CFOldStylePList.c) to check exactly what escapes it supports ```\``` escaped as ```\\``` is notable absent from the code listing on the blog post.

The only thing we would be unable to support is unpaired surrogates are these are not supported in UTF8 and therefore are not surpported by Rust's ```str``` or ```String```. As I can't imagine a reason for using them it's probably fine to error in that case.

**Comment by madig** [comment]
The test should probably be

```Rust
    #[test]
    fn escaped_sequences_in_strings() {
        let plist = r#"{
            key1 = "va\"lue";
            key2 = 'va"lue';
            key3 = "va\a\b\f\n\r\t\v\"\nlue";
            key4 = "a\012b";
            key5 = "\\UD83D\\UDCA9";
            key6 = "\UD83D\UDCA9";
        }"#;
        let cursor = Cursor::new(plist.as_bytes());
        let streaming_parser = AsciiReader::new(cursor);
        let events: Vec<Event> = streaming_parser.map(|e| e.unwrap()).collect();

        let comparison = &[
            StartDictionary(None),
            String("key1".to_owned()),
            String(r#"va"lue"#.to_owned()),
            String("key2".to_owned()),
            String(r#"va"lue"#.to_owned()),
            String("key3".to_owned()),
            String("va\u{07}\u{08}\u{0C}\n\r\t\u{0B}\"\nlue".to_owned()),
            String("key4".to_owned()),
            String("a\nb".to_owned()),
            String("key5".to_owned()),
            String("\\UD83D\\UDCA9".to_owned()),
            String("key6".to_owned()),
            String("💩".to_owned()),
            EndCollection,
        ];

        assert_eq!(events, comparison);
    }
```

Parses with openstep-plist:
```Python
{'key1': 'va"lue',
 'key2': 'va"lue',
 'key3': 'va\x07\x08\x0c\n\r\t\x0b"\nlue',
 'key4': 'a\nb',
 'key5': '\\UD83D\\UDCA9',
 'key6': '💩'}
```

Next step: implement the equivalent of https://github.com/opensource-apple/CF/blob/master/CFOldStylePList.c#L123-L171.

**Comment by anthrotype** [comment]
What is left for this PR to reach mergeable state, besides fixing the merge conflicts? Is @fstephany still interested in picking this one up?

**Comment by steven-joruk** [comment]
I'd like to see this feature land. @anthrotype @fstephany if you don't mind I'd be happy to finish it off, and if either of you have any branches with outstanding work I can take it from there, let me know.

**Comment by steven-joruk** [comment]
It didn't require many changes, here's my branch https://github.com/steven-joruk/rust-plist/commits/ascii-reader/

[The commit that fixes one of the serde tests](https://github.com/steven-joruk/rust-plist/commit/7c69512edee1085a0d9ee705bf0cad6d7a1f165e) highlighted that documents that don't begin with exactly `<?xml` will be treated as ascii, which might not be desirable.

The master branch is broken due to denying warnings and a deprecation, I updated to `swap_remove` in my branch.
```
error: use of deprecated method `indexmap::IndexMap::<K, V, S>::remove`: `remove` disrupts the map order -- use `swap_remove` or `shift_remove` for explicit behavior.
  --> src/dictionary.rs:70:18
   |
70 |         self.map.remove(key)
```

**Comment by steven-joruk** [comment]
My branch has been updated to include the escape sequence handling and the unicode mappings. It passes the tests provided by madig.

I had to allow escaping `\` (`\\`) because the test that parses `netnewswire.pbxproj` fails without it.

**Comment by pronebird** [comment]
It would be amazing to see this land. Especially because some of apple CLIs like launchctl return old style plists in stdout.

**Comment by ebarnard** [comment]
@steven-joruk If you can't push to this branch you should definitely open a new PR and I'll look at it afresh. I can't remember much of this one.

**Comment by ebarnard** [comment]
Closed via #136.
